### PR TITLE
Ignite-15603 : Calcite. NPE on subquery returning multiple results.

### PR DESCRIPTION
**Pull request by Vladsz83**
Fixed correlated pass in table_scan and index_scan in order to put exchange before correlate

### CODE REVIEW

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/LogicalScanConverterRule.java [hunk 1]
```diff
@@ -81,6 +81,12 @@
                     .replace(collation);
 
                 Set<CorrelationId> corrIds = RexUtils.extractCorrelationIds(rel.condition());
+
+                if (!F.isEmpty(rel.projects())) {
```
Comment (alex-plekhanov) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/LogicalScanConverterRule.java:85]: Why only for index scan? Table scan should be fixed too.

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/CalciteQueryProcessorTest.java [hunk 2]
```diff
@@ -1148,6 +1150,22 @@ public void testSelectWithOrdering() throws IgniteInterruptedCheckedException {
...
+    public void testSubqueryMultyValueError() throws IgniteInterruptedCheckedException {
+        sql("create table test_tbl(v INTEGER)", true);
+
+        sql("INSERT INTO test_tbl VALUES (1), (2), (3), (4), (5)", true);
+
+        Throwable err = GridTestUtils.assertThrows(log,
+            () -> sql("SELECT t0.v, (SELECT t0.v + t1.v FROM test_tbl t1) AS j FROM test_tbl t0"),
+            IgniteSQLException.class, null);
+
+        assertTrue(X.hasCause(err, "Subquery returned more than 1 value", IllegalArgumentException.class));
```
Comment (alex-plekhanov) [inline modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/CalciteQueryProcessorTest.java:1166]: Can we write a query that is not throwing an exception after a fix, but throwing an exception without fix? For example by reducing the number of rows in test_tbl? If not, perhaps we should write planner test.

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/CalciteQueryProcessorTest.java [hunk 3]
```diff
@@ -1148,6 +1150,22 @@ public void testSelectWithOrdering() throws IgniteInterruptedCheckedException {
             .check();
     }
 
+    /**
+     * Checks exchnage-before-correlate and correct multy-value subquery error.
+     */
+    @Test
+    public void testSubqueryMultyValueError() throws IgniteInterruptedCheckedException {
```
Comment (alex-plekhanov) [inline modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/CalciteQueryProcessorTest.java:1157]: The test name doesn't match javadoc. 
Actually, we should check correlated queries with correlates in projects (without correlates in filters), perhaps we should name the test (and fix the javadoc) according to this check. 
Also, fix typos (`multy` -> `multi`, `exchnage` -> `exchange`)

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/LogicalScanConverterRule.java (hunk 1 above)
Comment (Vladsz83) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/LogicalScanConverterRule.java:87]: Yep. IDE's Misscommit. Fixed.

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/CalciteQueryProcessorTest.java [hunk 4]
```diff
@@ -1148,6 +1150,18 @@ public void testSelectWithOrdering() throws IgniteInterruptedCheckedException {
...
+    /**
+     * Checks correlates are assigned before access.
+     */
+    @Test
+    public void testCorrelatesAssignedBeforeAccess() throws IgniteInterruptedCheckedException {
+        sql("create table test_tbl(v INTEGER)", true);
+
+        sql("INSERT INTO test_tbl VALUES (1)", true);
+
+        sql("SELECT t0.v, (SELECT t0.v + t1.v FROM test_tbl t1) AS j FROM test_tbl t0");
```
Comment (alex-plekhanov) [inline modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/CalciteQueryProcessorTest.java:1162]: Let's assert result here

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/CalciteQueryProcessorTest.java (hunk 3 above)
Comment (Vladsz83) [reply modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/CalciteQueryProcessorTest.java]: Spelling fixed. Test renamed.

Filters? No filter is supposed. The best approach, I believe, would be the tree analysis. Besides, the tree is one of major goals to evolve and optimize. Probably, more effective solution here is to watch NPE/exception not to appear.

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/CalciteQueryProcessorTest.java (hunk 4 above)
Comment (Vladsz83) [reply modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/CalciteQueryProcessorTest.java]: Done

**Review by alex-plekhanov — APPROVED** [review]

**Pull request closed without merging**